Enhance the backgrounds addon to support CSS variables.

konstantindenerz opened the pull request:
## Issue
In my case I have mono repo with different frontends. Each frontend should have own theme, but my core component should support all themes. In my stories I want to apply a specific theme. Therefore the change of the background color is not enough. Modern theming should use [CSS variables](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties). 

## What I did
Because of this I extended the background model of the backgrounds addon to support CSS variables. 

The are two ways to provide a theme:
1. Configuration
```js
  backgrounds: {
    default: 'theme-with-css-variables',
    values: [
      {
        name: 'theme-with-css-variables',
        value: '#ff4785',
        variables: {
          '--text-color': '#ff4785',
        },
      },
    ],
  },
```

2. Asset File (CSS)
```js
backgrounds: {
    default: 'theme-with-css-variables-from-asset-file',
    values: [
      {
        name: 'theme-with-css-variables-from-asset-file',
        value: '#ff4785',
        variablesAsset: 'assets/sample-theme-variables.css',
      },
    ],
  },
```
```css
/* sample-theme-variables.css */
:root {
  --text-color: #ff4785;
}
```

Depending on the parameters I create a style element or link element dynamically to apply CSS variables.

I renamed the tooltip and the clear button to make it clearer what is behind the action.


I think this should be a part of the essentials (addon bundle) and I think this should not be a separate addon to avoid confusion with background addon.

I need help if further documentation is necessary. For instance where is the right place to describe this?

## How to test

- Is this testable with Jest or Chromatic screenshots? yes
- Does this need a new example in the kitchen sink apps? yes
- Does this need an update to the documentation? yes

If your answer is yes to any of these, please make sure to include it in your PR.

---

I extended the existing story file for background addon stories:
`yarn storybook` in the `storybook/examples/official-storybook` folder
http://localhost:9011/?path=/story/addons-backgrounds--with-css-variables
http://localhost:9011/?path=/story/addons-backgrounds--with-css-variables-from-asset


<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `master` branch as part of the release process, so you shouldn't need to worry about this.

Maintainers: Please tag your pull request with at least one of the following:
`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

-->

### Screenshot
![image](https://user-images.githubusercontent.com/1221055/95013762-aa516000-0642-11eb-8408-c27f8a49ea13.png)




Please assign [**hacktoberfest-accepted**](https://hacktoberfest.digitalocean.com/) label to this PR.

konstantindenerz committed:
feat(backgrounds): apply CSS variables assigned to selected background from configuration or asset file

konstantindenerz committed:
feat(examples, official-storybook): add background stories to show CSS variables usage

yannbf commented:
Oh wow thanks a lot for your contribution @konstantindenerz! As soon as I have some time I will review this in more depth. Thanks!!

yannbf commented on addons/backgrounds/src/containers/BackgroundSelector.tsx:
```diff
@@ -131,8 +131,8 @@ export const BackgroundSelector: FunctionComponent = memo(() => {
           key="background"
-          title="Change the background of the preview"
-          active={selectedBackgroundColor !== 'transparent'}
+          title="Change the theme of the preview"
```
I'm a bit on the fence about this.. the addon handles backgrounds, and surely now it will be able to handle css variables, but still people can't use it to switch themes if not via css variables. This could confuse a few users that are using emotion/styled components or other solution that doesn't rely on css variables. What do you think @shilman ?

yannbf replied:
Also mind that there's a cypress test in `cypress/generated/addon-backgrounds.spec.ts` which will fail because of this title change.

shilman replied:
@yannbf I agree that this is still primarily an addon for changing backgrounds. Maybe the caption should be:

```
"Change the background or CSS of the preview"
```

yannbf commented on addons/backgrounds/src/helpers/index.ts:
```diff
@@ -5,22 +5,27 @@ import { logger } from '@storybook/client-logger';
 
-  if (backgrounds.find((background) => background.value === currentSelectedValue)) {
-    return currentSelectedValue;
+  const result = backgrounds.find((background) => background.value === currentSelectedValue);
+  if (result) {
+    return result;
```
```suggestion
  const isValueAlreadySelected = backgrounds.find((background) => background.value === currentSelectedValue);
  if (isValueAlreadySelected) {
    return currentSelectedValue;
```

yannbf reviewed (changes requested):
Hey @konstantindenerz I'm sorry it took me a long time to get back to you. This contribution is super nice and will help lots of people, thanks for doing this! Tried it locally and it works nicely, although the css variables will "bleed" in docs mode.. the last story will have precedence over the rest because its css variables will override the ones from the previous stories, but that's such an edge case that I think it's fine. 

I added just a small comment and a suggestion. 

It would be nice to have some explanation on how to use this feature in the docs though, could you do it? check `docs/essentials/backgrounds.md` as reference.

konstantindenerz commented:
Thank you for the review @yannbf! I can make a suggestion for the documentation part.

konstantindenerz committed:
refactor(backgrounds): Change visible texts to clarify the primary use case

konstantindenerz committed:
docs(backgrounds): add section for CSS variables

konstantindenerz commented:
Hey @yannbf , sorry for the delay. I updated the both labels, the tooltip and the button to clear. I added a new section in the docs to explain the new configuration options. Maybe you could review the changes again?
Do you think I should update the video too?

konstantindenerz committed:
test(addon-backgrounds): fix spec

yannbf commented:
> Hey @yannbf , sorry for the delay. I updated the both labels, the tooltip and the button to clear and I added a new section in the docs to explain the new configuration options. Maybe you could review the changes again?
> Do you think I should update the video too?

No worries! I'll review it as soon as I can!

darleendenno commented:
@yannbf are you able to review?

yannbf commented:
> @yannbf are you able to review?

Definitely, will do it today!

yannbf committed:
Merge branch 'next' into backgrounds-next

yannbf committed:
refactor(addon-backgrounds): rename css variables properties

yannbf reviewed (changes requested):
Hey @konstantindenerz I'm truly sorry for this huge delay! I updated this branch with latest from `next`, and made a small change to have properties clearly named to their purpose `cssVariables` and `cssVariablesAsset`, so people don't get confused.

On my side it looks good and works well! The only remarks I have are:
1 - Changes in a css asset that is imported locally won't trigger an update to the addon, but I think that is a limitation. Styles are updated if you switch stories back and forth. This could bring a bit of frustration to developers, so we might need to write something down in the documentation.
2 - If you have multiple stories that set the same css variables, if you open them in docs mode, the ones defined last will take precedence and override the rest. We can overcome this with scoped css variables applied when in docs mode

I think we should fix number 2, and I have no clue about number 1 (@ndelangen might have a clue). 

Let me know if you're still interested in continuing with the PR, otherwise I can implement a fix for #2. Thanks once again for your great contribution!

shilman commented:
Hi @konstantindenerz -- sorry for the long review process on this. A bunch of us discussed this at length as a group and we collectively that this extends the scope of a simple addon in ways we don't want to support (namely theming, which is a much broader problem). If you'd like to continue this work, we'd recommend you create a new addon with these changes and list it in the [Addon gallery](https://storybook.js.org/addons). Thank you again for your patience and contribution to storybook!
